File: behavioral/interpreter/query_language/python/query_language.py
from abc import ABC, abstractmethod
from typing import Dict, List, Any


class Expression(ABC):
    """Abstract base class for all expressions in our query language."""
    
    @abstractmethod
    def interpret(self, context: Dict[str, Any]) -> bool:
        """Interpret the expression against the given context."""
        pass


class LiteralExpression(Expression):
    """Terminal expression representing a literal value."""
    
    def __init__(self, value: Any):
        self.value = value
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        # Literals simply evaluate to themselves, but they're rarely used directly
        # This could be used for constant expressions like "TRUE" or "FALSE"
        return bool(self.value)


class VariableExpression(Expression):
    """Terminal expression representing a variable in the context."""
    
    def __init__(self, name: str):
        self.name = name
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        # Variables evaluate to their value in the context
        # This is primarily used internally by other expressions
        return bool(context.get(self.name, False))
    
    def get_value(self, context: Dict[str, Any]) -> Any:
        """Get the actual value from the context."""
        return context.get(self.name)


class EqualsExpression(Expression):
    """Non-terminal expression for equality comparison."""
    
    def __init__(self, variable: str, value: Any):
        self.variable = variable
        self.value = value
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        """Check if the variable equals the value, attempting type coercion."""
        context_value = context.get(self.variable)
        
        # Handle None comparison explicitly
        if context_value is None and self.value is None:
            return True
        if context_value is None or self.value is None:
            return False # One is None, the other isn't
        
        # Attempt type coercion for comparison
        try:
            # If context is numeric, try converting query value to that type
            if isinstance(context_value, (int, float)):
                if isinstance(self.value, (int, float)):
                    return context_value == self.value
                else:
                    return context_value == type(context_value)(self.value)
            # If query value is numeric, try converting context value to that type
            elif isinstance(self.value, (int, float)):
                 return type(self.value)(context_value) == self.value
            # Handle boolean comparison (case-insensitive string check)
            elif isinstance(context_value, bool):
                if isinstance(self.value, bool):
                    return context_value == self.value
                elif isinstance(self.value, str):
                    return context_value == (self.value.lower() == 'true')
            # Add more specific type comparisons if needed
        except (ValueError, TypeError):
            # Fallback to string comparison if coercion fails or types are incompatible
            pass 
            
        # Default/fallback: direct comparison (or string comparison)
        return str(context_value) == str(self.value)


class GreaterThanExpression(Expression):
    """Non-terminal expression for greater than comparison."""
    
    def __init__(self, variable: str, value: Any):
        self.variable = variable
        self.value = value
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        """Check if the variable is greater than the value, attempting numeric conversion."""
        context_value = context.get(self.variable)
        
        # Comparison only makes sense for numeric types usually
        try:
            # Ensure both can be treated as numbers (float for generality)
            num_context_value = float(context_value) 
            num_value = float(self.value)
            return num_context_value > num_value
        except (ValueError, TypeError, TypeError):
            return False # Cannot compare if not convertible to numbers


class LessThanExpression(Expression):
    """Non-terminal expression for less than comparison."""
    
    def __init__(self, variable: str, value: Any):
        self.variable = variable
        self.value = value
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        """Check if the variable is less than the value, attempting numeric conversion."""
        context_value = context.get(self.variable)
        
        # Comparison only makes sense for numeric types usually
        try:
            # Ensure both can be treated as numbers (float for generality)
            num_context_value = float(context_value)
            num_value = float(self.value)
            return num_context_value < num_value
        except (ValueError, TypeError):
            return False # Cannot compare if not convertible to numbers


class AndExpression(Expression):
    """Non-terminal expression for logical AND."""
    
    def __init__(self, left: Expression, right: Expression):
        self.left = left
        self.right = right
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        # Both expressions must be true
        return self.left.interpret(context) and self.right.interpret(context)


class OrExpression(Expression):
    """Non-terminal expression for logical OR."""
    
    def __init__(self, left: Expression, right: Expression):
        self.left = left
        self.right = right
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        # At least one expression must be true
        return self.left.interpret(context) or self.right.interpret(context)


class NotExpression(Expression):
    """Non-terminal expression for logical NOT."""
    
    def __init__(self, expression: Expression):
        self.expression = expression
    
    def interpret(self, context: Dict[str, Any]) -> bool:
        # Negate the result of the expression
        return not self.expression.interpret(context)


class QueryParser:
    """Parser for the query language, converting strings to expression trees."""
    
    def _find_split_index(self, query: str, operator: str) -> int:
        """Finds the index of the operator outside parentheses."""
        level = 0
        # Iterate checking for operator match at the current level (0)
        op_len = len(operator)
        for i in range(len(query) - op_len + 1):
            if query[i] == '(': level += 1
            elif query[i] == ')': level -= 1
            elif level == 0 and query[i:i+op_len] == operator:
                return i
        return -1

    def _try_convert_value(self, value_str: str) -> Any:
        """Tries to convert string value to int, float, bool, or keeps as string."""
        value_str = value_str.strip()
        # Remove potential quotes
        if (value_str.startswith('"') and value_str.endswith('"')) or \
           (value_str.startswith("'") and value_str.endswith("'")):
            return value_str[1:-1]
        # Check boolean
        if value_str.lower() == 'true': return True
        if value_str.lower() == 'false': return False
        # Check numeric
        try: return int(value_str)
        except ValueError:
            try: return float(value_str)
            except ValueError: return value_str # Keep as string

    def parse(self, query: str) -> Expression:
        """Parse a query string into an expression tree respecting precedence and parentheses."""
        query = query.strip()
        if not query: 
            raise ValueError("Cannot parse empty query")

        # 1. Handle parentheses if the entire query is enclosed
        if query.startswith('(') and query.endswith(')'):
            level = 0
            match = True
            # Check if the parentheses are matching outer ones
            for i, char in enumerate(query): 
                if char == '(': level += 1
                elif char == ')': level -= 1
                if level == 0 and i < len(query) - 1: # Closed before the end
                    match = False
                    break
            # Ensure level is 0 at the end and matching
            if match and level == 0: 
                return self.parse(query[1:-1])
            elif level != 0:
                 raise ValueError(f"Mismatched parentheses in query: {query}")
            # If not matching outer parentheses, proceed

        # 2. Handle OR (lowest precedence)
        or_index = self._find_split_index(query, ' OR ')
        if or_index != -1:
            left = self.parse(query[:or_index])
            right = self.parse(query[or_index + 4:]) # len(' OR ') == 4
            return OrExpression(left, right)

        # 3. Handle AND (next precedence)
        and_index = self._find_split_index(query, ' AND ')
        if and_index != -1:
            left = self.parse(query[:and_index])
            right = self.parse(query[and_index + 5:]) # len(' AND ') == 5
            return AndExpression(left, right)

        # 4. Handle NOT (prefix operator)
        if query.startswith('NOT '):
            expression = self.parse(query[4:])
            return NotExpression(expression)

        # 5. Handle comparison expressions (higher precedence)
        # Use find_split_index to avoid splitting inside potential nested function calls if added later
        comp_ops = [' = ', ' > ', ' < '] # Order matters if extending (e.g., >=)
        for op in comp_ops:
            op_index = self._find_split_index(query, op)
            if op_index != -1:
                variable = query[:op_index].strip()
                value_str = query[op_index + len(op):].strip()
                value = self._try_convert_value(value_str)
                
                if op == ' = ': return EqualsExpression(variable, value)
                if op == ' > ': return GreaterThanExpression(variable, value)
                if op == ' < ': return LessThanExpression(variable, value)
        
        # 6. Handle simple variable/literal (treat as variable for now)
        # Could add logic to detect quoted literals if needed
        # Check if it's a known boolean literal string just in case
        if query.lower() == 'true': return LiteralExpression(True)
        if query.lower() == 'false': return LiteralExpression(False)
        
        # Assume it's a variable name if not parsed otherwise
        # We could validate variable names here if needed (e.g., regex)
        return VariableExpression(query)


class QueryEngine:
    """Engine to run queries against collections of data."""
    
    def __init__(self):
        self.parser = QueryParser()
    
    def filter(self, data: List[Dict[str, Any]], query: str) -> List[Dict[str, Any]]:
        """Filter a list of data items using a query expression."""
        expression = self.parser.parse(query)
        return [item for item in data if expression.interpret(item)]

File: behavioral/interpreter/query_language/typescript/query_language.ts
/**
 * Interface for all expression types in the query language
 */
export interface Expression {
  interpret(context: Record<string, any>): boolean;
}

/**
 * Terminal expression representing a literal value
 */
export class LiteralExpression implements Expression {
  private value: any;

  constructor(value: any) {
    this.value = value;
  }

  interpret(_context: Record<string, any>): boolean {
    return Boolean(this.value);
  }
}

/**
 * Terminal expression representing a variable in the context
 */
export class VariableExpression implements Expression {
  private name: string;

  constructor(name: string) {
    this.name = name;
  }

  interpret(context: Record<string, any>): boolean {
    return Boolean(context[this.name]);
  }

  getValue(context: Record<string, any>): any {
    return context[this.name];
  }
}

/**
 * Non-terminal expression for equality comparison
 */
export class EqualsExpression implements Expression {
  private variable: string;
  private value: any;

  constructor(variable: string, value: any) {
    this.variable = variable;
    this.value = value;
  }

  interpret(context: Record<string, any>): boolean {
    const contextValue = context[this.variable];
    
    // Handle null/undefined comparison explicitly
    if (contextValue == null && this.value == null) {
      return true;
    }
    if (contextValue == null || this.value == null) {
      return false; // One is null/undefined, the other isn't
    }

    // Attempt type coercion for comparison
    try {
      // If context is numeric, try converting query value to that type
      if (typeof contextValue === 'number') {
        const numValue = Number(this.value);
        return !isNaN(numValue) && contextValue === numValue;
      } 
      // If query value is numeric, try converting context value to that type
      else if (typeof this.value === 'number') {
          const numContextValue = Number(contextValue);
          return !isNaN(numContextValue) && numContextValue === this.value;
      }
      // Handle boolean comparison (case-insensitive string check for query value)
      else if (typeof contextValue === 'boolean') {
        if (typeof this.value === 'boolean') {
          return contextValue === this.value;
        } else if (typeof this.value === 'string') {
          return contextValue === (this.value.toLowerCase() === 'true');
        }
      }
      // Add more specific type comparisons if needed
    } catch (e) {
      // Fallback to string comparison if coercion fails or types are incompatible
    }

    // Default/fallback: string comparison
    return String(contextValue) === String(this.value);
  }
}

/**
 * Non-terminal expression for greater than comparison
 */
export class GreaterThanExpression implements Expression {
  private variable: string;
  private value: any;

  constructor(variable: string, value: any) {
    this.variable = variable;
    this.value = value;
  }

  interpret(context: Record<string, any>): boolean {
    const contextValue = context[this.variable];
    
    // Comparison usually makes sense only for numeric types
    try {
      const numContextValue = Number(contextValue);
      const numValue = Number(this.value);
      // Check if both conversions were successful
      if (!isNaN(numContextValue) && !isNaN(numValue)) {
          return numContextValue > numValue;
      }
    } catch (e) { 
      // Ignore conversion errors
    }
    return false; // Cannot compare if not convertible to numbers
  }
}

/**
 * Non-terminal expression for less than comparison
 */
export class LessThanExpression implements Expression {
  private variable: string;
  private value: any;

  constructor(variable: string, value: any) {
    this.variable = variable;
    this.value = value;
  }

  interpret(context: Record<string, any>): boolean {
    const contextValue = context[this.variable];

    // Comparison usually makes sense only for numeric types
    try {
      const numContextValue = Number(contextValue);
      const numValue = Number(this.value);
      // Check if both conversions were successful
      if (!isNaN(numContextValue) && !isNaN(numValue)) {
          return numContextValue < numValue;
      }
    } catch (e) {
      // Ignore conversion errors
    }
    return false; // Cannot compare if not convertible to numbers
  }
}

/**
 * Non-terminal expression for logical AND
 */
export class AndExpression implements Expression {
  private left: Expression;
  private right: Expression;

  constructor(left: Expression, right: Expression) {
    this.left = left;
    this.right = right;
  }

  interpret(context: Record<string, any>): boolean {
    return this.left.interpret(context) && this.right.interpret(context);
  }
}

/**
 * Non-terminal expression for logical OR
 */
export class OrExpression implements Expression {
  private left: Expression;
  private right: Expression;

  constructor(left: Expression, right: Expression) {
    this.left = left;
    this.right = right;
  }

  interpret(context: Record<string, any>): boolean {
    return this.left.interpret(context) || this.right.interpret(context);
  }
}

/**
 * Non-terminal expression for logical NOT
 */
export class NotExpression implements Expression {
  private expression: Expression;

  constructor(expression: Expression) {
    this.expression = expression;
  }

  interpret(context: Record<string, any>): boolean {
    return !this.expression.interpret(context);
  }
}

/**
 * Parser for the query language, converting strings to expression trees
 */
export class QueryParser {

  private findSplitIndex(query: string, operator: string): number {
    let level = 0;
    const opLen = operator.length;
    for (let i = 0; i <= query.length - opLen; i++) {
      if (query[i] === '(') {
        level++;
      } else if (query[i] === ')') {
        level--;
      } else if (level === 0 && query.substring(i, i + opLen) === operator) {
        return i;
      }
    }
    return -1;
  }

  private tryConvertValue(valueStr: string): any {
    valueStr = valueStr.trim();
    // Remove potential quotes
    if ((valueStr.startsWith('"') && valueStr.endsWith('"')) ||
        (valueStr.startsWith("'") && valueStr.endsWith("'"))) {
      valueStr = valueStr.substring(1, valueStr.length - 1);
    }
    // Check boolean
    if (valueStr.toLowerCase() === 'true') return true;
    if (valueStr.toLowerCase() === 'false') return false;
    // Check numeric
    const numVal = Number(valueStr);
    if (!isNaN(numVal)) return numVal;
    
    return valueStr; // Keep as string
  }

  parse(query: string): Expression {
    query = query.trim();
    if (!query) {
      // Handle empty query - returning Literal(false) or throwing error
      return new LiteralExpression(false); // Placeholder
    }

    // 1. Handle parentheses if the entire query is enclosed
    if (query.startsWith('(') && query.endsWith(')')) {
        let level = 0;
        let match = true;
        // Check if the parentheses are matching outer ones
        for (let i = 0; i < query.length; i++) {
            const char = query[i];
            if (char === '(') level++;
            else if (char === ')') level--;
            if (level === 0 && i < query.length - 1) { // Closed before the end
                match = false;
                break;
            }
        }
        // Ensure level is 0 at the end and matching
        if (match && level === 0) {
            return this.parse(query.substring(1, query.length - 1));
        } else if (level !== 0) {
             // Handle mismatched parentheses - returning Literal(false) or throwing error
             return new LiteralExpression(false); // Placeholder
        }
        // If not matching outer parentheses, proceed
    }

    // 2. Handle OR (lowest precedence)
    const orIndex = this.findSplitIndex(query, ' OR ');
    if (orIndex !== -1) {
        const left = this.parse(query.substring(0, orIndex));
        const right = this.parse(query.substring(orIndex + 4)); // length of ' OR '
        return new OrExpression(left, right);
    }

    // 3. Handle AND (next precedence)
    const andIndex = this.findSplitIndex(query, ' AND ');
    if (andIndex !== -1) {
        const left = this.parse(query.substring(0, andIndex));
        const right = this.parse(query.substring(andIndex + 5)); // length of ' AND '
        return new AndExpression(left, right);
    }

    // 4. Handle NOT (prefix operator)
    if (query.startsWith('NOT ')) {
      const expression = this.parse(query.substring(4));
      return new NotExpression(expression);
    }

    // 5. Handle comparison expressions
    const compOps = [' = ', ' > ', ' < ']; // Order matters if extending
    for (const op of compOps) {
        const opIndex = this.findSplitIndex(query, op);
        if (opIndex !== -1) {
            const variable = query.substring(0, opIndex).trim();
            const valueStr = query.substring(opIndex + op.length).trim();
            const value = this.tryConvertValue(valueStr);

            switch (op) {
                case ' = ': return new EqualsExpression(variable, value);
                case ' > ': return new GreaterThanExpression(variable, value);
                case ' < ': return new LessThanExpression(variable, value);
            }
        }
    }

    // 6. Handle simple variable/literal
    // Check if it's a known boolean literal string just in case
    if (query.toLowerCase() === 'true') return new LiteralExpression(true);
    if (query.toLowerCase() === 'false') return new LiteralExpression(false);

    // Assume it's a variable name if not parsed otherwise
    return new VariableExpression(query);
  }
}

/**
 * Engine to run queries against collections of data
 */
export class QueryEngine {
  private parser: QueryParser;

  constructor() {
    this.parser = new QueryParser();
  }

  filter(data: Record<string, any>[], query: string): Record<string, any>[] {
    const expression = this.parser.parse(query);
    return data.filter(item => expression.interpret(item));
  }
}
